### ISSUE TITLE
Gateway Quiz: Answers Available Setting Changes per Student?

### ISSUE THREAD

**Issue by thatchermo** [openwebwork/webwork2#1109]
I ran into behavior that I found surprising in WebWork 2.15, and that I think might be a bug--at least to this end user. I created a Gateway Quiz set to be open Tuesday through Friday, with answers available to students the following Monday so that I'd have time to review all their submissions and possibly add partial credit. However, while looking at student scores on their completed versions of the test, I noticed that some values had been rewritten in the settings for the version that was created for each student. Screenshots attached below (I hope).

It makes sense to me that Open and Close in the actual versions of the test change to reflect when the student took the test, but note that the Answers Available blank also changes, and for the student who took the test very early, the answers are actually available before the test closes for other students. This seems like a bug to me, given that it could allow the early test-taker to see answers and communicate them to later test-takers (obviously partially mitigated by randomization, but not everything is randomized).

I've certainly done nothing on purpose to cause it to behave like that, but would be curious if I've done something by accident.

I know it's bad form to report two bugs in one issue, but possibly related is the fact that the early test-taker did reclick the link for the exam to see her scores while the test was still open for others, and it recorded new blank answers for all questions. Previous answers were still recorded, and I could access them by clicking the "Previous Answers" button, and her total score remained the same, but seeing all answers blank is obviously worrysome for her and me. I haven't been able to recreate exactly what she did (and students aren't always good about telling you everything they might have clicked), but I wonder if this behavior is related to the answers available behavior.

![ww_class](https://user-images.githubusercontent.com/16291971/81580275-da862280-9372-11ea-9ee0-f92143015082.png)
![ww_student1](https://user-images.githubusercontent.com/16291971/81580276-db1eb900-9372-11ea-8f36-59f6d484a5bf.png)
![ww_student2](https://user-images.githubusercontent.com/16291971/81580279-dbb74f80-9372-11ea-8a95-c0be020470bc.png)

**Comment by pschan-gh** [comment]
We experienced the same issue quoted below. All problems on a student's quiz recorded blank answers after the student had reportedly clicked on the link to the quiz under "Homework Sets" after / around the submission deadline. The score of their quiz was still computed according to their last actually submitted answers though.

The settings for our quiz are as follows:

![Screenshot 2021-01-21 at 11 08 52 AM](https://user-images.githubusercontent.com/31397850/105274871-14354b80-5bd9-11eb-8253-c22a8a805bd6.png)

> 
> I know it's bad form to report two bugs in one issue, but possibly related is the fact that the early test-taker did reclick the link for the exam to see her scores while the test was still open for others, and it recorded new blank answers for all questions. Previous answers were still recorded, and I could access them by clicking the "Previous Answers" button, and her total score remained the same, but seeing all answers blank is obviously worrysome for her and me. I haven't been able to recreate exactly what she did (and students aren't always good about telling you everything they might have clicked), but I wonder if this behavior is related to the answers available behavior.
>

**Comment by drlvk** [comment]
Indeed, WeBWorK generates individualized "answer date" values for each student's attempt, which is different from "template answer date" which you set.  It uses the formula 
(template answer date) - (template due date) + (student's due date) 
where student's due date is usually (their starting time + time limit). See below. 

https://github.com/openwebwork/webwork2/blob/8f8bff9757b913a5ecce65fff533cf3ae49861ed/lib/WeBWorK/ContentGenerator/GatewayQuiz.pm#L761-L771

However, the above computation has no effect on the actual availability of answers or question scores, which is still determined by the "template answer date". 

https://github.com/openwebwork/webwork2/blob/8f8bff9757b913a5ecce65fff533cf3ae49861ed/lib/WeBWorK/ContentGenerator/GatewayQuiz.pm#L106-L108

The discrepancy is confusing to students because they are shown the individualized "answer date" if they return to the quiz before the answers are available. The day/time they see there is **not** when the answer will actually become available.

https://github.com/openwebwork/webwork2/blob/8f8bff9757b913a5ecce65fff533cf3ae49861ed/lib/WeBWorK/ContentGenerator/GatewayQuiz.pm#L106-L108

It seems to me that not creating individualized "answer date" would be the right approach.

**Comment by drlvk** [comment]
> there are at least some institutions who expect it to work this way.

But it does not actually work this way at present. WeBWorK computes individualized "answer date" for each student but does not use it.  It uses the "template answer date" when deciding whether to show the answers. 

For example, right now (09/17/2021 02:38pm EDT) I logged in as one of students in a course and see the message 

> Completed results for this assignment are not available until 09/16/2021 at 06:40pm EDT

which is a contradiction to itself: this "individualized answer date" is well in the past. 

In reality, the results will become available at the Answer Date set globally by the instructor.

**Comment by somiaj** [comment]
Just adding I can confirm the above issue with the individual answer dates vs set answer dates. Students aren't allowed to access the quiz even though the displayed individual answer date is in the past while the set answer date is in the future. So either this is not working as intended or it is printing an incorrect date.

I'll second the idea of a configurable option to allow students to see quizzes before the set answer date or not.
